### ISSUE TITLE
Model arguments aren't translated into actual classes

### ISSUE THREAD

**Issue by stsievert** [adriangb/scikeras#82]
Let's say I use this code:

``` python
est = KerasRegressor(
    model=model,
    model__foo=Foo,  # user defined class
    model__foo__bar=Bar,  # user defined class
    model__foo__bar__x=5,
)
```

I would expect `model(foo=Foo(Bar(x=5)))` to be called to build the model. Instead, I get a ValueError because `model` doesn't accept argument `foo__bar`.

Complete example:

<details>

``` python
import tensorflow as tf
from scikeras.wrappers import KerasRegressor

class Foo:
    def __init__(self, bar):
        self.bar = bar

class Bar:
    def __init__(self, x=3):
        self.x = x

def model(foo: Foo):
    inputs = tf.keras.Input(shape=(100,))
    x = tf.keras.layers.Dense(4, activation=tf.nn.relu)(inputs)
    outputs = tf.keras.layers.Dense(1, activation=tf.nn.softmax)(x)
    model = tf.keras.Model(inputs=inputs, outputs=outputs)
    return model

est = KerasRegressor(
    model=model,
    model__foo=Foo,
    model__foo__bar=Bar,
    model__foo__bar__x=5,
)
```

</details>

**Comment by adriangb** [comment]
Why would you expect that behavior? I don't think we ever suggested support for that (nor do I think we should). Maybe we can make the utilities we use to route parameters and compile optimizers/losses/metrics available to users so that they can invoke them within `model` and achieve the behavior your are looking for?

```python3
def model(params):
    foo = params["foo"]
    f = foo(**route_params(params, destination="foo")
```

But I don't think we should, by default, "compile" arbitrary user parameters.

**Comment by stsievert** [comment]
> Why would you expect that behavior?

Because the `loss`, `optimizer` and `metrics` support the same behavior. I don't see why the behavior is inconsistent depending on the keyword.

I do not expect this to be a common use case. Thinking about it more, I don't know if it's worth more than a documentation warning.

**Comment by adriangb** [comment]
I think the fundamental difference is that those are parameters to `Model.compile` and are hardcoded in `BaseWrapper.__init__` with documentation of this behavior. `foo` is a user defined parameter.

Do you think that this might be a use case to justify making these utilities public?

**Comment by adriangb** [comment]
> Do you think that this might be a use case to justify making these utilities public?

@stsievert any thoughts on this? If we are going to make these public, I'd rather do that before working on #79. It should be straightforward (just rename and maybe add back the tests we ended up removing in #66)

**Comment by stsievert** [comment]
I don't think the utilities (`route_params`,`parse_params`) be public, certainly not with the current signatures.

**Comment by adriangb** [comment]
Is it because the current signatures are not clear, or not clean? The use case above looks pretty useful to users.

**Comment by stsievert** [comment]
> The use case above looks pretty useful to users.

Then why not unflatten parameters that begin with`model__`?

That said, I don't think it's super useful. In https://github.com/adriangb/scikeras/issues/82#issuecomment-698578194, where would `foo` be used? It could only be used in building the model; `model.compile` already has unflattened arguments.

> Is it because the current signatures are not clear, or not clean?

I think because the interface is confusing. I have these questions on the current signatures:

* Why does `unflatten_params` require two parameters? What's the third optional parameter used for?
* I can see why `route_params` has the first two required parameters. Why is `pass_filter` required? What does `strict` do?

**Comment by adriangb** [comment]
> Then why not unflatten parameters that begin with`model__`?

Because users may or may not want that behavior. If they're the ones calling `unflatten`, then there will be no unexpected behavior.

> That said, I don't think it's super useful.

If you don't think this would be useful then let's drop it and close this issue.

> I think because the interface is confusing.

Then let's fix it (at some point). I think that all of the things you point out can either be refactored out or explained clearly in 1-2 lines of a docstring.
